src/valkey-cli.c

```diff
@@ -2224,6 +2226,29 @@ static int cliReadReply(int output_raw_strings) {
         fflush(stdout);
         sdsfree(out);
     }
+
+    /* Handle pubsub mode */
+    if (config.pubsub_mode) {
+        if (isPubsubPush(reply)) {
+            if (reply->elements >= 3) {
+                char *cmd = reply->element[0]->str;
+                int count = reply->element[2]->integer;
+
+                if (strcmp(cmd, "subscribe") == 0 || strcmp(cmd, "unsubscribe") == 0 ||
+                    strcmp(cmd, "psubscribe") == 0 || strcmp(cmd, "punsubscribe") == 0) {
+                    config.pubsub_unsharded_count = count;
+                } else if (strcmp(cmd, "ssubscribe") == 0 || strcmp(cmd, "sunsubscribe") == 0) {
+                    config.pubsub_sharded_count = count;
+                }
+                
+                if (config.pubsub_unsharded_count == 0 && config.pubsub_sharded_count == 0) {
+                    config.pubsub_mode = 0;
+                    cliRefreshPrompt();
+                }
+            }
+        }
+    }
```

Comment:
The logic looks right to me. I think we should move it though.

You have put this logic in `cliReadReply`, but I think we should put it where `cliReadReply` is called, just a few lines below. That's where most of the client state is handled and it's also where the logic is that sets `config.pubsub_mode = 1;`, so it seems more logical to put this in the same place. The reply can be accessed as `config.last_reply`.

Somewhere here:

```C

        /* Read response, possibly skipping pubsub/push messages. */
        while (1) {
            if (cliReadReply(output_raw) != REDIS_OK) {
                zfree(argvlen);
                return REDIS_ERR;
            }
            fflush(stdout);
            if (config.pubsub_mode || num_expected_pubsub_push > 0) {
                if (isPubsubPush(config.last_reply)) {
                    if (num_expected_pubsub_push > 0 && !strcasecmp(config.last_reply->element[0]->str, command)) {
                        /* This pushed message confirms the
                         * [p|s][un]subscribe command. */
                        if (is_subscribe && !config.pubsub_mode) {
                            config.pubsub_mode = 1;
                            cliRefreshPrompt();
                        }
```